cmd/gcp-controller-manager/kubelet_readonly_csr_approver.go

```diff
@@ -252,31 +253,67 @@ func validatePodAnnotation(request kubeletReadonlyCSRRequest) kubeletReadonlyCSR
 		klog.Errorf("csr does not have an attached pod name")
 		return kubeletReadonlyCSRResponse{
 			result:  false,
-			err:     fmt.Errorf("csr does not have an attached pod name"),
+			err:     nil,
 			message: "csr does not have pod name attached",
 		}
 	}
 
+	podUIDs, ok := request.csr.Spec.Extra[podUIDKey]
+	if !ok {
+		klog.Errorf("csr does not have an attached pod UID")
+		return kubeletReadonlyCSRResponse{
+			result:  false,
+			err:     nil,
+			message: "csr does not have pod UID attached",
+		}
+	}
+
+	if len(podNames) != len(podUIDs) {
+		return kubeletReadonlyCSRResponse{
+			result:  false,
+			err:     nil,
+			message: "bad request: length of pod name and UID are not equal",
+		}
+	}
+
 	for _, podName := range podNames {
-		pod, err := request.controllerContext.client.CoreV1().Pods("").Get(request.context, podName, metav1.GetOptions{})
+		klog.Infof("start finding pod %s", podName)
+		pods, err := request.controllerContext.client.CoreV1().Pods("").List(request.context, metav1.ListOptions{
```

Comment:
this has performance implications at scale, @marseel @wojtek-t can you please chime in?
should we set `resourceVersion=0`?

Reply:
IIRC a List has always to retrieve all the list from etcd to filter it in the apiserver, so forcing it to be served from the cache alleviates the problem

Reply:
Do you have any better way to get a pod with uid but has no namespace?

Reply:
Yes, you are right, it may cause scale issue. 

Another way I can think is to list all namespace. but it will cause the same problem. for the scale issue, i don't think it will be a problem. Since we only have few CSR for each cluster(for now only datadog, metric server, Splunk).

Reply:
well, if you already have a pod informer I think you can use the local cache to look for the pod.uid, but let's wait for Marcel or Wojtek advice, they are more used to these problems

Reply:
Cool, cache may help

Reply:
So what you're trying to achieve is basically list pods with a given name from all namespaces, right?

Assuming that - listing like you did is a terrible idea, because it will basically list *all* pods from etcd, deserialize them, filter out and only then return the couple you're interested in.

Setting `RV=0` would help, but still we're not indexing pods by name in cache (only by nodeName:
https://github.com/kubernetes/kubernetes/blob/master/pkg/registry/core/pod/strategy.go#L277
), so we would still need to iterate over all of them and filter out.


IIRC, gcp-controller already has a pod-informer at a higher-level, so we should instead propagate it down to here, and list pods from the informer (ideally adding appropriate index to make that efficient).

Reply:
But the question is, usually the csr will be made during pod initiation stage, so the cache will have large chance missed.

Reply:
actually, I check code, when we have directPath, it will add a nodesyncer into the map, and the map is sorted by UID, but in my case, it is not guarantee that directPath is enabled. I have two option
1. enable node syncer no matter  direct Path enable or not. I haven't look deep into the code, it may cost more effort
2. Rewrite a syncer only for pod pool and let node syncer reuse the code. 

What do you think

Reply:
https://github.com/kubernetes/cloud-provider-gcp/blob/043a2d5bab9959aa503806a099637049fe78d436/cmd/gcp-controller-manager/loops.go#L127

here is the link I mentioned

And I have one question, what about I list all namespaces and get pod for each namespace. it will reduce a lot?